truffle/src/main/java/org/jruby/truffle/runtime/methods/SharedMethodInfo.java

```diff
@@ -35,6 +48,59 @@ public SharedMethodInfo(SourceSection sourceSection, LexicalScope lexicalScope,
         this.isBlock = isBlock;
         this.parseTree = parseTree;
         this.alwaysSplit = alwaysSplit;
+        this.keywordArguments = null;
+        this.arity = null;
+    }
+    
+    public SharedMethodInfo(SourceSection sourceSection,
+            LexicalScope lexicalScope, String name, boolean isBlock,
+            org.jruby.ast.Node parseTree, boolean alwaysSplit, ArgsNode argsNode) {
+        assert sourceSection != null;
+        assert name != null;
+
+        this.sourceSection = sourceSection;
+        this.lexicalScope = lexicalScope;
+        this.name = name;
+        this.isBlock = isBlock;
+        this.parseTree = parseTree;
+        this.alwaysSplit = alwaysSplit;
+
+        if (argsNode.hasKwargs()) {
+            keywordArguments = new ArrayList<String>();
+            if (argsNode.getKeywords() != null) {
+                for (Node node : argsNode.getKeywords().childNodes()) {
+                    final KeywordArgNode kwarg = (KeywordArgNode) node;
+                    final AssignableNode assignableNode = kwarg.getAssignable();
+
+                    if (assignableNode instanceof LocalAsgnNode) {
+                        keywordArguments.add(((LocalAsgnNode) assignableNode)
+                                .getName());
+                    } else if (assignableNode instanceof DAsgnNode) {
+                        keywordArguments.add(((DAsgnNode) assignableNode)
+                                .getName());
+                    } else {
+                        throw new UnsupportedOperationException(
+                                "unsupported keyword arg " + node);
+                    }
+                }
+            }
+        } else {
+            keywordArguments = null;
+        }
+        
+        this.arity = getArity(argsNode);
```

Comment:
Also not sure whether this is the correct way to do it: we added arity here so that we can access it in the dispatch node (get number of required args etc. to determine whether we should optimize/expand a keywords hash. If the last argument is a Hash but we are short on named arguments, we have to treat the Hash as a regular last argument instead of using it for keyword args. That case is not optimized.)